src/Utilities/Compiler/Options/SymbolNamesWithValueOption.cs

```diff
@@ -43,89 +81,287 @@ private SymbolNamesWithValueOption()
+                    return true;
+                }
+
+                symbolKindsToCheck.Add(SymbolKind.NamedType);
+                currentType = currentType.ContainingType;
+            }
+
+            // ...Now try a partial check looping on the symbol's containing namespace...
+            INamespaceSymbol? currentNamespace = symbol.ContainingNamespace;
+            while (currentNamespace != null && !currentNamespace.IsGlobalNamespace)
+            {
+                if (TryGetSymbolPartialMatch(symbolFullNameBuilder, currentNamespace, out firstMatch))
+                {
+                    return true;
+                }
+
+                symbolKindsToCheck.Add(SymbolKind.Namespace);
+                currentNamespace = currentNamespace.ContainingNamespace;
+            }
+
+            // ...At this point we couldn't match any part of the symbol name in the 'AllKinds' part of the list, let's try with the type fully qualified name...
+            Debug.Assert(symbolFullNameBuilder.Length > 0);
+            Debug.Assert(symbolKindsToCheck.Count >= 1 && symbolKindsToCheck.Count <= 3);
+
+            var symbolFullName = symbolFullNameBuilder.ToString();
+
+            foreach (var kind in symbolKindsToCheck)
+            {
+                if (!_wildcardNamesBySymbolKind.ContainsKey(kind))
+                {
+                    continue;
+                }
+
+                if (TryGetFirstWildcardMatch(kind, symbolFullName, out firstMatch))
+                {
+                    return true;
+                }
+            }
+
+            // ...No match
+            Debug.Assert(firstMatch.Equals(NoWildcardMatch));
+            _wildcardMatchResult.AddOrUpdate(symbol, NoWildcardMatch, (s, kvp) => NoWildcardMatch);
+            return false;
+
+            // Local functions.
+            bool TryGetSymbolPartialMatch(StringBuilder builder, ISymbol symbol, out KeyValuePair<string, TValue> firstMatch)
+            {
+                if (builder.Length > 0)
+                {
+                    builder.Insert(0, ".");
+                }
+
+                builder.Insert(0, symbol.Name);
+
+                return TryGetFirstWildcardMatch(AllKinds, symbol.Name, out firstMatch);
+            }
+
+            bool TryGetFirstWildcardMatch(SymbolKind kind, string symbolName, out KeyValuePair<string, TValue> firstMatch)
+            {
+                if (!_wildcardNamesBySymbolKind.ContainsKey(kind))
+                {
+                    firstMatch = NoWildcardMatch;
+                    return false;
+                }
+
+                firstMatch = _wildcardNamesBySymbolKind[kind].FirstOrDefault(x => symbolName.StartsWith(x.Key, StringComparison.Ordinal));
+
+                if (string.IsNullOrWhiteSpace(firstMatch.Key))
+                {
+                    return false;
+                }
+
+                var match = firstMatch;
+                _wildcardMatchResult.AddOrUpdate(symbol, firstMatch, (s, kvp) => match);
```

Comment:
This method is another example where it is not clear to me when and why cache is updated and why is it not updated on the the false returning paths.